proposals/NNNN-dxil-vectors.md

```diff
@@ -0,0 +1,183 @@
+<!-- {% raw %} -->
+
+# DXIL Vectors
+
+---
+
+* Proposal: [NNNN](NNNN-dxil-vectors.md)
+* Author(s): [Greg Roth](https://github.com/pow2clk)
+* Sponsor: [Greg Roth](https://github.com/pow2clk)
+* Status: **Under Consideration**
+* Planned Version: Shader Model 6.9
+
+## Introduction
+
+While DXIL is intended and able to support language vectors,
+ those vectors must be broken up into individual scalars to be valid DXIL.
+This feature introduces the ability to represent native vectors in DXIL for some uses.
+
+## Motivation
+
+While the original shape of the vectors may be reconstructed from their scalarized form,
+ it requires additional work of the DXIL consumer and results in larger DXIL binary sizes.
+Although it has never been allowed in DXIL, the LLVM IR that DXIL is based on can represent native vectors.
+By allowing these native vector types in DXIL, the size of generated DXIL can be reduced and
+ new opportunities for expanding vector capabilities in DXIL ar introduced.
+
+## Proposed solution
+
+Native vectors are allowed in DXIL version 1.9 or greater.
+These can be stored in allocas, static globals, and groupshared variables.
+They can be loaded from or stored to raw buffers and used as arguments to a selection
+ of element-wise intrinsic functions as well as the standard math operators.
+They cannot be used in shader signatures, constant buffers, typed buffer, or texture types.
+
+## Detailed design
+
+### Vectors in memory representations
+
+In their alloca and variable representations, vectors in DXIL will always be represented as vectors.
+Previously individual vectors would get scalarized into scalar arrays and arrays of vectors would be flattened
+ into a one-dimensional scalar array with indexing to reflect the original intents.
+Individual vectors will now be represented as a single native vector and arrays of vectors will remain
+ as arrays of native vectors, though multi-dimensional arrays will still be flattened to one dimension.
+
+Scalarization of these vectors will continue to be done for uses that don't support native vectors,
+ but it will be done using extractelement instructions from the native vectors
+ instead of loads from the scalarized array representation.
+
+Single-element vectors are not valid in DXIL.
+At the language level, they may be supported for corresponding intrinsic overloads,
+  but such vectors should be represented as scalars in the final DXIL output.
```

Comment:
It'd be good to understand the rationale for this.  Is this just about reducing testing / implementation cost, or is there something more fundamental I'm missing?

If I'm following correctly, this would make it pretty hard to implement something like specialization constants in DXIL where the constant controlled the size of the vector.

Reply:
I guess this is something we'll have to tackle when we start thinking about adding native specialization constants to HLSL.